proposals/0001-dependency-injection.md

```diff
@@ -0,0 +1,633 @@
+integrated dependency injection system.
+
+### Core Requirements
+
+1. Usability
+    - It must have a simple interface and be accessible for new users.
+    - It must support named dependencies to resolve ambiguities.
+    - It must avoid temporal coupling when declaring dependencies (i.e., declaration order should not matter).
+    - It should work "out of the box" without introducing unnecessary complexity.
+2. Performance
+    - It must be performant, avoiding runtime bottlenecks such as excessive reflection or repeated heavy computations.
+3. Modularity
+    - The system must be optional and seamlessly integrate into the existing Ktor platform.
+    - It must be extensible, so developers can adjust or override its behavior as needed.
+
+### Type support
+
+1. Support for generic types, interfaces, and type variance (e.g., covariant types during resolution).
+2. Ability to instantiate types using reflection as an optional feature.
+
+### Tooling
+
+1. The system must support runtime references for flexibility in dependency selection.
+2. It should allow dependency declaration and configuration via external files (e.g., `application.conf`).
+3. It should avoid any reliance on annotations or meta-programming, staying consistent with Ktor's "no magic" philosophy.
+4. It should permit compile-time validation of dependencies to avoid runtime errors from missing dependencies.
+
+# Design Details
+[design-details]: #design-details
+
+We'll break down the API by the different phases of its usage:
+
+1. **Declaration:** how the instantiation of different types is registered
+2. **Resolution:** how we can access the different types from the running application
+
+When developing a Ktor application, you can expect it to look something like this:
+
+```kotlin
+fun main() {
+   embeddedServer(Netty, port = 8080, host = "0.0.0.0", module = Application::module)
+      .start(wait = true)
+}
+
+fun Application.module() {
+   declareDependencies()    // See "Declaration" section 
+   configureApplication()   // See "Resolution" section
+}
+```
+
+In this example, both modules are called from the main `module()`, but they can be treated as interchangeable units 
+using configuration.
+
+For example, when using YAML configuration: 
+
+```yaml
+# application.yaml / ktor / application 
+modules:
+   - io.ktor.example.ImplementationKt.declareDependencies
+   - io.ktor.example.ApplicationKt.configureApplication
+```
+
+This way, we can introduce a dependency inversion where the `declareDependencies` module can live on a different runtime
+classpath than the application module `configureApplication`, which has no knowledge of the implementation details.
+
+In the next two sections, we'll provide the API design for both declaring and resolving dependencies.
+
+## Declaration
+[declaration]: #declaration
+
+As with declaring extensions in a Ktor application, users can expect to declare their dependencies from the
+same scope: `Application`.
```

Comment:
Don't you consider any other scope besides the main default "Application" one? No request scope definitions?